unify http ratelimiter for control plane client

zach593 opened the pull request:
**What type of PR is this?**

/kind bug


<!--
Add one of the following kinds:

/kind api-change
/kind bug
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

#6094

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
For all components with rate limit parameters of 'kube-api-qps' and 'kube-api-burst', and scheduler-estimator with 'cluster-api-qps' and 'cluster-api-burst', these parameters will be stricter and administrators may need to adjust them to get the same effect as before the update.
```

XiShanYongYe-Chang commented on pkg/util/ratelimiter.go:
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2024 The Karmada Authors.
```
2024 -> 2025

zach593 replied:
done.

XiShanYongYe-Chang commented on pkg/util/ratelimiter.go:
```diff
@@ -0,0 +1,69 @@
+limitations under the License.
+*/
+
+package util
```
Can you help add some ut for this file?

zach593 replied:
updated.

zach593 commented:
> Thanks a lot~
> 
> The PR looks a little big. Can it be split, for example, cluster or karmada, or different components?
> 
> In addition, are there any modified test results?

Ah, I missed this comment. Sure, I'll split this commit ASAP.

As for the test results... Of course I could post a picture showing  20 qps to karmada and 5 qps to cluster, but I don't find that convincing, but testing qps/burst as UT is a bit hard, let me see what I can do about it.

zach593 commented:
@XiShanYongYe-Chang I just updated this PR to remove the member cluster ratelimiter part. The remaining part is simple and easy to review.

For uniform usage, I replaced all qps/burst in rest.Config with ratelimiter.

I'll go ahead and do the rest.

XiShanYongYe-Chang reviewed (commented):
Thanks~
/lgtm

zach593 commented:
the second part is separated to #6192

zach593 commented:
> Hi @zach593, do you think we need to cherry-pick this patch to the previous release branchs?

This doesn't seem like something that needs to be fixed urgently, and users have been using it in its previous state for too long to realize it's a problem, so I don't think it needs to be cherry-picked.

RainbowMango commented:
> This doesn't seem like something that needs to be fixed urgently, and users have been using it in its previous state for too long to realize it's a problem, so I don't think it needs to be cherry-picked.

+1

XiShanYongYe-Chang commented:
Thanks for your reply~
#6192 should also be consistent logic.

Let's merge this pr first.
/approve

RainbowMango commented:
> For all components with rate limit parameters of 'kube-api-qps' and 'kube-api-burst', and scheduler-estimator with 'cluster-api-qps' and 'cluster-api-burst', these parameters will be stricter and administrators may need to adjust them to get the same effect as before the update.

I feel hard to understand the release notes. And 

> administrators may need to adjust them to get the same effect as before the update.

If I'm the administrator, I don't know how to adjust these flags.

XiShanYongYe-Chang commented:
How about this release note?
```
Unify the rate limiter for different clients in each component to access karmada-apiserver. Access may be restricted in large-scale environments compared to before the modification. Administrators can avoid this situation by adjusting the qps and bursts for each component to access karmada-apiserver.
```

zhzhuang-zju commented:
> Administrators can avoid this situation by adjusting the qps and bursts for each component to access karmada-apiserver.

How about `Administrators can avoid this situation by adjusting upward the rate limit parameters 'kube-api-qps' and 'kube-api-burst' for each component to access karmada-apiserver.`?

zach593 commented:
> Unify the rate limiter for different clients in each component to access karmada-apiserver. Access may be restricted in large-scale environments compared to before the modification. Administrators can avoid this situation by adjusting upward the rate limit parameters 'kube-api-qps' and 'kube-api-burst' for each component to access karmada-apiserver,  and adjusting 'cluster-api-qps' and 'cluster-api-burst' for scheduler-estimator to access member cluster apiserver. 

There are also some changes to the estimator, which are added and become like this. Is this ok? It seems a bit long.

XiShanYongYe-Chang commented:
> There are also some changes to the estimator, which are added and become like this. Is this ok? It seems a bit long.

/cc @zhzhuang-zju 
Is it convenient to add some more information?

zhzhuang-zju commented:
> There are also some changes to the estimator, which are added and become like this. Is this ok? It seems a bit long.

Thanks, it looks much clearer. The length doesn't matter. As long as the meaning is expressed clearly, it's okay.
